### ISSUE TITLE
`torch.clamp` does not distribute gradients as element-wise`min/max` do

### ISSUE THREAD

**Issue by takagi** [pytorch/pytorch#76522]
### 🐛 Describe the bug

Since #59669 was merged, element-wise `max/min` evenly distribute gradients for all values that are equal in self and the other. However, `clump`, which can be equivalently expressed using `man` and `min`, does not follow the way, resulting in different gradients between `clamp` and `min/max`. I suppose the similar fix with #59669 should be also applied to `clamp` so that it distributes gradients as `min/max` do.

```py
>>> import torch
>>> a = torch.tensor([-2.,-1.,0.,1.,2.], requires_grad=True)
>>> b = torch.full((5,), -1.); b.requires_grad = True
>>> c = torch.full((5,), 1.); c.requires_grad = True
>>> torch.clamp(a, b,c).sum().backward()
>>> a.grad, b.grad, c.grad
(tensor([0., 1., 1., 1., 0.]), tensor([1., 0., 0., 0., 0.]), tensor([0., 0., 0., 0., 1.]))
>>> a.grad, b.grad, c.grad = None, None, None
>>> torch.min(torch.max(a, b), c).sum().backward()
>>> a.grad, b.grad, c.grad
(tensor([0.0000, 0.5000, 1.0000, 0.5000, 0.0000]), tensor([1.0000, 0.5000, 0.0000, 0.0000, 0.0000]), tensor([0.0000, 0.0000, 0.0000, 0.5000, 1.0000]))
```

### Versions

Collecting environment information...
PyTorch version: 1.11.0+cpu
Is debug build: False
CUDA used to build PyTorch: Could not collect
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.6 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: 6.0.0-1ubuntu2 (tags/RELEASE_600/final)
CMake version: version 3.10.2
Libc version: glibc-2.27

Python version: 3.10.0 (default, Oct 15 2021, 11:40:42) [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-4.15.0-169-generic-x86_64-with-glibc2.27
Is CUDA available: False
CUDA runtime version: Could not collect
GPU models and configuration: Could not collect

Nvidia driver version: Could not collect
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.21.2
[pip3] torch==1.11.0+cpu

**Comment by Lezcano** [comment]
Makes sense to me. wdyt @albanD ?

**Comment by Sentient07** [comment]
This issue affects me as well. Commenting to follow

**Comment by albanD** [comment]
Yes we should make that change.

**Comment by albanD** [comment]
To add a bit more details:
- The overload of the function that takes 3 Tensors should spread the gradients as suggested (the fact that some inputs might not require gradients should be ignored as only the differentiability is important here).
- The overload that takes numbers as input is a bit different as the numbers are not differentiable. In this case, we have locally a convex (resp concave) function and so we can take the min-norm sub (resp super) gradient which will give us 0.

Does that sound good?

**Comment by Lezcano** [comment]
For reference, this approach sgtm. We should document it properly when we implement it, as it's a tricky one.

**Comment by vadimkantorov** [comment]
> * actually clamp doesn't do Number -> Tensor promotion

It does broadcast/promote PyTorch scalars https://github.com/pytorch/pytorch/issues/71918, but not Python scalars :(

**Comment by ezyang** [comment]
What if we YOLO turned on Python scalar to wrapped number Tensor arguments for more functions (right now only add/mul/div and a few others are special cased for this)

**Comment by Lezcano** [comment]
See https://github.com/pytorch/pytorch/issues/76555#issuecomment-1112620338 for @rgommers on this topic

**Comment by albanD** [comment]
Note that in this case, we are actually interested in seeing the difference because you will get two different formulas for the two overloads!

**Comment by ILoveSE** [comment]
In the 1.13.1 and 2.0(dev) nightly versions, I can still observe these different gradients with `torch.Tensor.clamp` and `torch.clamp`.
Is this a problem or just an expected behavior?
I wonder whether this behavior will be fixed in future versions.




```python
import torch

torch.manual_seed(0)
a = torch.tensor([-2.,-1.,0.,1.,2.], requires_grad=True)
b = torch.full((5,), -1.)
b.requires_grad = True
c = torch.full((5,), 1.)
c.requires_grad = True

torch.Tensor.clamp(a, b, c).sum().backward()
# torch.clamp(a, b, c).sum().backward()
x1=a.grad
y1=b.grad
z1=c.grad
print(x1,y1,z1)

a.grad, b.grad, c.grad = None, None, None
torch.min(torch.max(a, b), c).sum().backward()
x2=a.grad
y2=b.grad
z2=c.grad


print(x2,y2,z2)
```

<details>
  <summary>pytorch 1.13.1</summary>
  <pre><code>
  [pip3] numpy==1.22.3
  [pip3] torch==1.13.1
  [pip3] torchelastic==0.2.2
  [pip3] torchtext==0.14.1
  [pip3] torchvision==0.14.1
  [conda] blas                      1.0                         mkl  
  [conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
  [conda] mkl                       2021.4.0           h06a4308_640  
  [conda] mkl-service               2.4.0           py310h7f8727e_0  
  [conda] mkl_fft                   1.3.1           py310hd6ae3a3_0  
  [conda] mkl_random                1.2.2           py310h00e6091_0  
  [conda] numpy                     1.22.3          py310hfa59a62_0  
  [conda] numpy-base                1.22.3          py310h9585f30_0  
  [conda] pytorch                   1.13.1          py3.10_cuda11.6_cudnn8.3.2_0    pytorch
  [conda] pytorch-cuda              11.6                 h867d48c_1    pytorch
  [conda] pytorch-mutex             1.0                        cuda    pytorch
  [conda] torchelastic              0.2.2                    pypi_0    pypi
  [conda] torchtext                 0.14.1                    py310    pytorch
  [conda] torchvision               0.14.1              py310_cu116    pytorch</code></pre>
</details>

<details>
  <summary>pytorch nightly</summary>
  <pre><code>
  [pip3] numpy==1.11.3
  [pip3] torch==2.0.0.dev20230219
  [pip3] torchaudio==2.0.0.dev20230219
  [pip3] torchvision==0.15.0.dev20230219
  [conda] blas                      1.0                         mkl  
  [conda] mkl                       2020.2                      256  
  [conda] mkl-service               2.3.0            py38he904b0f_0  
  [conda] numpy                     1.11.3          py38h7e9f1db_12  
  [conda] numpy-base                1.11.3          py38hde5b4d6_12  
  [conda] pytorch                   2.0.0.dev20230219 py3.8_cuda11.7_cudnn8.5.0_0    pytorch-nightly
  [conda] pytorch-cuda              11.7                 h778d358_3    pytorch-nightly
  [conda] pytorch-mutex             1.0                        cuda    pytorch-nightly
  [conda] torchaudio                2.0.0.dev20230219      py38_cu117    pytorch-nightly
  [conda] torchtriton               2.0.0+c8bfe3f548            py38    pytorch-nightly
  [conda] torchvision               0.15.0.dev20230219      py38_cu117    pytorch-nightly</code></pre>
</details>
